File: smallpict-rust/tests/models_test.rs
use smallpict::models::{
    FitMode, ImageFormat, OptimizeOptions, PurgeResponse, QuotaResponse,
};

#[test]
fn test_optimize_options_builder() {
    let opts = OptimizeOptions::builder()
        .format(ImageFormat::Avif)
        .quality(85)
        .max_width(1920)
        .max_height(1080)
        .fit(FitMode::Contain)
        .lossless(false)
        .strip_metadata(true)
        .filename("banner.png")
        .mime_type("image/png")
        .idempotency_key("idemp_123")
        .build();

    assert_eq!(opts.format, Some(ImageFormat::Avif));
    assert_eq!(opts.quality, Some(85));
    assert_eq!(opts.max_width, Some(1920));
    assert_eq!(opts.max_height, Some(1080));
    assert_eq!(opts.fit, Some(FitMode::Contain));
    assert_eq!(opts.filename, Some("banner.png".to_string()));
    assert_eq!(opts.idempotency_key, Some("idemp_123".to_string()));

    let json_str = serde_json::to_string(&opts).expect("valid serialization");
    assert!(json_str.contains("\"format\":\"avif\""));
    assert!(json_str.contains("\"quality\":85"));
    assert!(!json_str.contains("idempotency_key")); // skipped from JSON body
}

#[test]
fn test_serde_models() {
    let quota_json = r#"{
        "plan": "api_velocity",
        "bytes_used": 5000000,
        "quota_limit": 10000000,
        "quota_percentage": 50.0,
        "active_keys_count": 3
    }"#;
    let quota: QuotaResponse = serde_json::from_str(quota_json).expect("valid quota deserialize");
    assert_eq!(quota.plan, "api_velocity");
    assert_eq!(quota.bytes_used, 5000000);
    assert_eq!(quota.active_keys_count, Some(3));

    let purge_json = r#"{"message":"Purge completed"}"#;
    let purge: PurgeResponse = serde_json::from_str(purge_json).expect("valid purge deserialize");
    assert_eq!(purge.message, "Purge completed");
}

File: smallpict-java/src/test/java/com/smallpict/ModelsTest.java
package com.smallpict;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.smallpict.models.*;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ModelsTest {
    private final ObjectMapper mapper = new ObjectMapper();

    @Test
    void testOptimizeOptionsBuilder() throws Exception {
        OptimizeOptions options = OptimizeOptions.builder()
                .format(ImageFormat.AVIF)
                .quality(85)
                .maxWidth(1920)
                .maxHeight(1080)
                .fit(FitMode.CONTAIN)
                .lossless(false)
                .stripMetadata(true)
                .filename("banner.png")
                .mimeType("image/png")
                .idempotencyKey("idemp_123")
                .build();

        assertEquals(ImageFormat.AVIF, options.getFormat());
        assertEquals(85, options.getQuality());
        assertEquals(1920, options.getMaxWidth());
        assertEquals(FitMode.CONTAIN, options.getFit());
        assertEquals("banner.png", options.getFilename());
        assertEquals("idemp_123", options.getIdempotencyKey());

        String json = mapper.writeValueAsString(options);
        assertTrue(json.contains("\"format\":\"avif\""));
        assertTrue(json.contains("\"quality\":85"));
        assertFalse(json.contains("idempotencyKey")); // ignored in json body
    }

    @Test
    void testOptimizeResultDeserialization() throws Exception {
        String json = """
        {
            "job_id": "job_123",
            "status": "completed",
            "url": "https://cdn.smallpict.app/opt/hero.avif",
            "format": "avif",
            "original_size": 100000,
            "compressed_size": 15000,
            "bytes_saved": 85000,
            "savings_percentage": 85.0
        }
        """;

        OptimizeResult result = mapper.readValue(json, OptimizeResult.class);
        assertEquals("job_123", result.getJobId());
        assertEquals("completed", result.getStatus());
        assertEquals("https://cdn.smallpict.app/opt/hero.avif", result.getUrl());
        assertEquals(85.0, result.getSavingsPercentage());
        assertEquals(85000, result.getBytesSaved());
    }

    @Test
    void testQuotaResponseDeserialization() throws Exception {
        String json = """
        {
            "plan": "api_velocity",
            "bytes_used": 5000000,
            "quota_limit": 10000000,
            "quota_percentage": 50.0,
            "active_keys_count": 3
        }
        """;

        QuotaResponse quota = mapper.readValue(json, QuotaResponse.class);
        assertEquals("api_velocity", quota.getPlan());
        assertEquals(5000000L, quota.getBytesUsed());
        assertEquals(10000000L, quota.getQuotaLimit());
        assertEquals(50.0, quota.getQuotaPercentage());
        assertEquals(3, quota.getActiveKeysCount());
    }
}
